twine/commands/upload.py

```diff
@@ -48,6 +50,40 @@ def skip_upload(response, skip_existing, package):
             (response.status_code == 403 and msg_403 in response.text)))
 
 
+def check_status_code(response, verbose):
+    """
+    Additional safety net to catch response code 410 in case the
+    UploadToDeprecatedPyPIDetected exception breaks.
+    Also includes a check for response code 405 and prints helpful error
+    message guiding users to the right repository endpoints.
+    """
+    if response.status_code == 410 and "pypi.python.org" in response.url:
+        raise exceptions.UploadToDeprecatedPyPIDetected(
+            f"It appears you're uploading to pypi.python.org (or "
+            f"testpypi.python.org). You've received a 410 error response. "
+            f"Uploading to those sites is deprecated. The new sites are "
+            f"pypi.org and test.pypi.org. Try using {DEFAULT_REPOSITORY} (or "
+            f"{TEST_REPOSITORY}) to upload your packages instead. These are "
+            f"the default URLs for Twine now. More at "
+            f"https://packaging.python.org/guides/migrating-to-pypi-org/.")
+    elif response.status_code == 405 and "pypi.org" in response.url:
+        raise exceptions.InvalidPyPIUploadURL(
```

Comment:
Per your other suggestion:

> How do you feel about raising a single exception for `InvalidPyPIUploadURL` and `UploadToDeprecatedPyPIDetected` as they're both in the realm of invalid URLs, and passing in relevant messages when raising them?

I like it, though I might take it further and define the exception as:

```python
class InvalidPyPIUploadURL(TwineException):
    def __init__(self):
        return super().__init__(
            "It appears you're trying to upload to PyPI (or TestPyPI) but "
            "have an invalid URL. You probably want one of these two URLs: "
            f"{DEFAULT_REPOSITORY} or {TEST_REPOSITORY}. "
            "Check your repository configuration."
        )
```

Which feels like it encompasses both scenarios, removes duplicate message strings, and could be used simply as:

```python
raise exceptions.InvalidPyPIUploadURL()
```

However, that requires some more substantial code changes to replace in `UploadToDeprecatedPyPIDetected`. It also loses the link to the migration doc, and I don't know how important that is.

With that in mind, I'm going to approve this (with the small suggestion of using the language above for this exception), and request a review from the other @pypa/twine-maintainers.

Reply:
Sounds good, and I'd be down to work on replacing those exceptions. I don't see any significant number of replacements either. :)

Reply:
@di and @sigmavirus24: As the folks who seem to be most familiar with the interaction of Twine and Warehouse, I'd appreciate your feedback on this. It looks like there's an opportunity to simplify/refactor some of the exceptions, but I'm curious if it's okay to increase the scope of this PR to do that.

Reply:
Let's do that in a separate PR